File: crates/printobserver-server/tests/support/world.rs
//! What every journey here drives: a real server, over a real HTTP client.
//!
//! Nothing below stands in for the layer under test. The router, the policy,
//! the store, the ingress and the vision adapter are the real ones; the two
//! things a journey supplies are the external systems on the far side of a
//! port — the machine and the supervising agent — and the real ones are what
//! this crate's integration tier runs against.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};
use std::sync::Arc;

use printobserver_obico::{ObicoVision, ObicoVisionConfig};
use printobserver_server::{Ports, Running, Server, ServerConfig};
use printobserver_store_api::StorePort;
use printobserver_store_sqlite::SqliteStore;
use printobserver_types::{PrintId, serde_json};
use tempfile::TempDir;

use crate::agent::StandInAgent;
use crate::printer::RecordingPrinter;

/// The shared secret every journey's ingress requires.
pub const SECRET: &str = "a-shared-secret-nothing-else-knows";

/// The file the configuration is written to under a journey's own root.
pub const CONFIG_FILE: &str = "config.toml";

/// One configuration document, in the shape the file takes.
///
/// Built from values rather than from a string, so a journey that makes one
/// field unacceptable changes that field and nothing else.
#[must_use]
pub fn document(root: &Path, octoprint_url: &str) -> toml::Value {
    let text = format!(
        r#"
state_dir = "{state}"
listen = "127.0.0.1:0"

[octoprint]
url = "{octoprint_url}"
api_key = "a-provisioned-key"
fan = "commandable"

[supervisor]
harness = "claude-code"

[ingress]
shared_secret = "{SECRET}"
answer_bound_ms = 1000

[safety]
agent_min_interval_s = 0

[safety.allowed]
feedrate = {{ min = 0.5, max = 1.5 }}
flowrate = {{ min = 0.9, max = 1.1 }}
fan = {{ min = 0.0, max = 100.0 }}
bed_target = {{ min = 0.0, max = 110.0 }}
"tool_target:0" = {{ min = 0.0, max = 260.0 }}

[safety.actions]
operator = ["pause", "resume", "cancel", "start_print", "set_feedrate_factor",
            "set_flowrate_factor", "set_tool_target_c", "set_bed_target_c",
            "set_fan_percent", "acknowledge_failure"]
agent = ["pause", "set_feedrate_factor", "set_fan_percent"]
system = ["set_feedrate_factor", "set_flowrate_factor", "set_tool_target_c",
          "set_bed_target_c", "set_fan_percent"]
"#,
        state = root.join("state").display(),
    );
    toml::from_str(&text).expect("the base configuration is a document")
}

/// Set one dotted key of a configuration document.
///
/// # Panics
///
/// Panics when the path names no table, which is a journey naming a field the
/// document does not have.
pub fn set(document: &mut toml::Value, dotted: &str, value: toml::Value) {
    let (table, last) = table_of(document, dotted);
    table.insert(last, value);
}

/// Remove one dotted key of a configuration document.
///
/// # Panics
///
/// Panics when the path names no table.
pub fn remove(document: &mut toml::Value, dotted: &str) {
    let (table, last) = table_of(document, dotted);
    table.remove(&last);
}

/// The table one dotted key lives in, and the key's own last segment.
fn table_of<'a>(
    document: &'a mut toml::Value,
    dotted: &str,
) -> (&'a mut toml::Table, String) {
    let mut segments: Vec<&str> = dotted.split('.').collect();
    let last = segments.pop().expect("a key has at least one segment").to_owned();
    let mut here = document;
    for segment in segments {
        here = here
            .get_mut(segment)
            .unwrap_or_else(|| panic!("the document has no `{segment}` table"));
    }
    (
        here.as_table_mut()
            .unwrap_or_else(|| panic!("`{dotted}` does not name a table entry")),
        last,
    )
}

/// Write one configuration document into a root, and answer its path.
pub fn write(root: &Path, document: &toml::Value) -> PathBuf {
    let path = root.join(CONFIG_FILE);
    std::fs::write(
        &path,
        toml::to_string(document).expect("a configuration document renders"),
    )
    .expect("the configuration is writable");
    path
}

/// One journey's own world: a running server, and what it was composed from.
pub struct World {
    /// The journey's root, removed when it is dropped.
    pub root: TempDir,
    /// The server, serving.
    pub server: Running,
    /// The machine it is acting on.
    pub printer: Arc<RecordingPrinter>,
    /// The agent it is supervising through.
    pub agent: Arc<StandInAgent>,
    /// Durable state, as the server holds it.
    pub store: Arc<dyn StorePort>,
    /// The client a journey drives the real surface with.
    pub client: reqwest::Client,
}

impl World {
    /// A server over a fresh root, with a machine part-way through a print.
    pub async fn open() -> Self {
        Self::open_with(RecordingPrinter::printing(), StandInAgent::new()).await
    }

    /// A server over a fresh root, over the machine and agent given.
    pub async fn open_with(
        printer: Arc<RecordingPrinter>,
        agent: Arc<StandInAgent>,
    ) -> Self {
        let root = TempDir::new().expect("a journey's own root");
        let path = write(root.path(), &document(root.path(), "http://127.0.0.1:1"));
        let config = ServerConfig::load(&path).expect("the base configuration is accepted");
        let (server, store) = start(&config, &printer, &agent).await;
        Self {
            root,
            server,
            printer,
            agent,
            store,
            client: reqwest::Client::new(),
        }
    }

    /// Stop this server and start another over the same state directory.
    ///
    /// This is what a restart is: the store and the session directory are the
    /// whole of the state, and nothing else is carried across.
    pub async fn restart(self) -> Self {
        let Self {
            root,
            server,
            printer,
            agent,
            client,
            ..
        } = self;
        let config = server.config().clone();
        server.stop().await;
        let (server, store) = start(&config, &printer, &agent).await;
        Self {
            root,
            server,
            printer,
            agent,
            store,
            client,
        }
    }

    /// The state directory this world's store lives in.
    #[must_use]
    pub fn state_dir(&self) -> PathBuf {
        self.server.config().state_dir.clone()
    }

    /// One URL beneath the versioned prefix.
    #[must_use]
    pub fn url(&self, path: &str) -> String {
        format!("{}{path}", self.server.base_url())
    }

    /// One whole path on this server, prefix included.
    #[must_use]
    pub fn at(&self, whole_path: &str) -> String {
        format!("http://{}{whole_path}", self.server.address())
    }

    /// One operation's URL, with the print substituted into it.
    ///
    /// The path an operation declares already carries the versioned prefix, so
    /// this is taken as a whole path rather than beneath one.
    #[must_use]
    pub fn operation_url(&self, path: &str, print_id: PrintId) -> String {
        self.at(&path.replace("{print_id}", &print_id.to_string()))
    }

    /// Open one print to act on.
    pub async fn open_print(&self) -> PrintId {
        self.store
            .open_print(Some(4211), Some("benchy.gcode".to_owned()))
            .await
            .expect("a print opens")
            .id
    }

    /// Read one answer without asking it to be JSON, for a journey about a
    /// path this server serves nothing at.
    pub async fn raw_get(&self, url: &str) -> (reqwest::StatusCode, String, String) {
        let response = self
            .client
            .get(url)
            .send()
            .await
            .expect("the server answers");
        let status = response.status();
        let media = response
            .headers()
            .get(reqwest::header::CONTENT_TYPE)
            .and_then(|value| value.to_str().ok())
            .unwrap_or_default()
            .to_owned();
        let text = response.text().await.expect("an answer has a body");
        (status, media, text)
    }

    /// Read one JSON answer, and the status it came under.
    pub async fn get(&self, url: &str) -> (reqwest::StatusCode, serde_json::Value) {
        let response = self
            .client
            .get(url)
            .header(reqwest::header::ACCEPT, printobserver_server::MEDIA_TYPE)
            .send()
            .await
            .expect("the server answers");
        read(response).await
    }

    /// Send one JSON body, and read the JSON answer.
    pub async fn post(
        &self,
        url: &str,
        body: &serde_json::Value,
    ) -> (reqwest::StatusCode, serde_json::Value) {
        let response = self
            .client
            .post(url)
            .json(body)
            .send()
            .await
            .expect("the server answers");
        read(response).await
    }

    /// Replace one record with a JSON body, and read the JSON answer.
    pub async fn put(
        &self,
        url: &str,
        body: &serde_json::Value,
    ) -> (reqwest::StatusCode, serde_json::Value) {
        let response = self
            .client
            .put(url)
            .json(body)
            .send()
            .await
            .expect("the server answers");
        read(response).await
    }
}

/// The status, the media type and the body of one answer.
async fn read(response: reqwest::Response) -> (reqwest::StatusCode, serde_json::Value) {
    let status = response.status();
    let media = response
        .headers()
        .get(reqwest::header::CONTENT_TYPE)
        .and_then(|value| value.to_str().ok())
        .unwrap_or_default()
        .to_owned();
    let text = response.text().await.expect("an answer has a body");
    assert!(
        media.starts_with(printobserver_server::MEDIA_TYPE),
        "the server answered under {media:?} rather than JSON: {text}"
    );
    let value = serde_json::from_str(&text)
        .unwrap_or_else(|error| panic!("the answer is not JSON ({error}): {text}"));
    (status, value)
}

/// Start one server over the ports a journey supplies.
async fn start(
    config: &ServerConfig,
    printer: &Arc<RecordingPrinter>,
    agent: &Arc<StandInAgent>,
) -> (Running, Arc<dyn StorePort>) {
    let store: Arc<dyn StorePort> = Arc::new(
        SqliteStore::open(&config.state_dir).expect("the store opens on the state directory"),
    );
    let vision =
        Arc::new(ObicoVision::new(ObicoVisionConfig::default()).expect("the adapter is built"));
    let server = Server::start_with(
        config.clone(),
        Ports {
            printer: Arc::clone(printer) as Arc<dyn printobserver_printer_api::PrinterPort>,
            store: Arc::clone(&store),
            vision,
            agent: Arc::clone(agent) as Arc<dyn printobserver_supervisor_api::SupervisorPort>,
        },
    )
    .await
    .expect("the server starts");
    (server, store)
}

/// One `Obico` failure alert, as the producer posts it.
///
/// The committed sample is what the ingress journeys post; this is what a
/// journey that needs a *second* alert about the same print, or an alert naming
/// an image host it controls, builds from it.
#[must_use]
pub fn failure_alert(obico_print_id: i64, image_url: &str) -> serde_json::Value {
    let mut sample: serde_json::Value =
        serde_json::from_str(committed_sample()).expect("the committed sample is JSON");
    sample["print"]["id"] = serde_json::json!(obico_print_id);
    sample["img_url"] = serde_json::json!(image_url);
    sample
}

/// The committed `Obico` failure alert sample, read from the contracts' own file.
#[must_use]
pub fn committed_sample() -> &'static str {
    include_str!("../../../printobserver-types/samples/obico/failure-alert.json")
}

/// Every event kind one print's history carries, counted.
#[must_use]
pub fn kinds(events: &serde_json::Value) -> BTreeMap<String, usize> {
    let mut found = BTreeMap::new();
    for event in events.as_array().into_iter().flatten() {
        let kind = event["kind"].as_str().unwrap_or_default().to_owned();
        *found.entry(kind).or_insert(0) += 1;
    }
    found
}

File: tools/release-artifacts/src/release_artifacts/world.py
"""A real supervisor to prove an installed client against.

A smoke check that reached no server would prove nothing about the artifact it
runs from, so each of the three runs against a **real `printobserver server`**:
the program this repository builds, started under a configuration this writes,
with a print and an image in its store to read.

Two things stand beside it and neither is inside what is being proven. The
printer is an `OctoPrint` stand-in on a real socket — the supervisor reaches it
over HTTP exactly as it reaches the machine beside the printer, and what proves
the same operations against a real `OctoPrint` is the printer integration tier.
The print and the image are opened by posting one `Obico` failure alert to the
supervisor's own ingress, which is how a print comes to exist at all; their
identifiers are then read out of the store the supervisor wrote.
"""

from __future__ import annotations

import json
import sqlite3
import subprocess
import threading
import time
from dataclasses import dataclass
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from types import TracebackType

from repo_checks.shell import start

#: The shared word the ingress requires of every post.
INGRESS_WORD = "a-shared-word-for-a-smoke-check"

#: The header that word travels in, as the ingress spells it.
INGRESS_HEADER = "x-printobserver-token"

#: The file the supervisor writes the address it bound into, for the clients
#: beside it. Reading it is how this finds a server started on a port the
#: operating system chose.
CLIENT_CONFIG = "client.toml"

#: The store the supervisor keeps its record in.
STORE = "printobserver.sqlite3"

#: A one-pixel image, so the alert this posts carries a snapshot the supervisor
#: really fetches and really stores.
IMAGE = bytes.fromhex(
    "ffd8ffe000104a46494600010101006000600000ffdb0043000806060706"
    "05080707070909080a0c140d0c0b0b0c1912130f141d1a1f1e1d1a1c1c20"
    "242e2720222c231c1c2837292c30313434341f27393d38323c2e333432ff"
    "c0000b080001000101011100ffc400140001000000000000000000000000"
    "00000009ffc40014100100000000000000000000000000000000ffda0008"
    "010100003f002a9fffd9"
)

#: The statuses the ingress answers a post it took under.
ACCEPTED = (200, 202)

#: How long the supervisor is given to answer at the address it bound.
STARTUP_TIMEOUT_SECONDS = 60.0

#: How long the ingress is given to open the print its alert names.
INGRESS_TIMEOUT_SECONDS = 60.0


@dataclass(frozen=True, slots=True)
class Running:
    """A supervisor a client can be pointed at."""

    #: Where it answers, as its own client configuration writes it.
    server: str
    #: The print every read of the smoke checks is about.
    print_id: str
    #: The image the materialization read is about.
    image_id: str
    #: Where its record and its images live.
    state: Path


class WorldError(RuntimeError):
    """A supervisor could not be brought up for a client to be proven against."""


class Machine:
    """An `OctoPrint` the supervisor reaches over HTTP, holding one printing job."""

    def __init__(self) -> None:
        """Start answering on a port the operating system chooses."""
        self.asked: list[str] = []
        self._server = ThreadingHTTPServer(("127.0.0.1", 0), _machine_handler(self))
        self._serving = threading.Thread(target=self._server.serve_forever, daemon=True)
        self._serving.start()

    @property
    def url(self) -> str:
        """Where this machine answers."""
        host, port = self._server.server_address[:2]
        return f"http://{host!s}:{port}"

    def stop(self) -> None:
        """Stop answering, leaving no thread behind."""
        self._server.shutdown()
        self._server.server_close()
        self._serving.join(timeout=10)

    def answer(self, path: str) -> tuple[int, bytes]:
        """What this machine answers one read with."""
        self.asked.append(path)
        if path.startswith("/api/printer"):
            return 200, json.dumps(
                {
                    "state": {
                        "text": "Printing",
                        "flags": {"operational": True, "printing": True},
                    },
                    "temperature": {
                        "tool0": {"actual": 210.0, "target": 210.0, "offset": 0.0},
                        "bed": {"actual": 60.0, "target": 60.0, "offset": 0.0},
                    },
                }
            ).encode()
        if path.startswith("/api/job"):
            return 200, json.dumps(
                {
                    "state": "Printing",
                    "job": {
                        "file": {"name": "benchy.gcode", "origin": "local", "size": 4096},
                        "estimatedPrintTime": 3600.0,
                    },
                    "progress": {
                        "completion": 12.5,
                        "printTime": 450,
                        "printTimeLeft": 3150,
                    },
                }
            ).encode()
        if path.startswith("/snapshot"):
            return 200, IMAGE
        return 204, b""


def _machine_handler(machine: Machine) -> type[BaseHTTPRequestHandler]:
    """The request handler the stand-in machine answers through."""

    class Handler(BaseHTTPRequestHandler):
        """Answer a read, and take a write without moving anything."""

        protocol_version = "HTTP/1.1"

        def _answer(self, status: int, body: bytes) -> None:
            self.send_response(status)
            self.send_header(
                "Content-Type",
                "image/jpeg" if body[:2] == b"\xff\xd8" else "application/json",
            )
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)

        def do_GET(self) -> None:
            """Answer a read."""
            status, body = machine.answer(self.path)
            self._answer(status, body)

        def do_POST(self) -> None:
            """Take a write, recording that it arrived."""
            length = int(self.headers.get("Content-Length") or 0)
            self.rfile.read(length)
            machine.asked.append(self.path)
            self._answer(204, b"")

        def log_message(self, format: str, *args: object) -> None:
            """Say nothing: a journey's output is its assertions, not its traffic."""

    return Handler


def _configuration(state: Path, machine: str) -> str:
    """The one configuration file the supervisor reads, as a document."""
    return json.dumps(
        {
            "state_dir": str(state),
            "listen": "127.0.0.1:0",
            "octoprint": {"url": machine, "api_key": "a-provisioned-key", "fan": "commandable"},
            "supervisor": {"harness": "claude-code"},
            "ingress": {"shared_secret": INGRESS_WORD, "answer_bound_ms": 1000},
            "safety": {
                "agent_min_interval_s": 0,
                "allowed": {
                    "feedrate": {"min": 0.5, "max": 1.5},
                    "flowrate": {"min": 0.9, "max": 1.1},
                    "fan": {"min": 0.0, "max": 100.0},
                    "bed_target": {"min": 0.0, "max": 110.0},
                    "tool_target:0": {"min": 0.0, "max": 260.0},
                },
                "actions": {
                    "operator": [
                        "pause",
                        "resume",
                        "cancel",
                        "start_print",
                        "set_feedrate_factor",
                        "set_flowrate_factor",
                        "set_tool_target_c",
                        "set_bed_target_c",
                        "set_fan_percent",
                        "acknowledge_failure",
                    ],
                    "agent": ["pause", "set_feedrate_factor"],
                    "system": ["pause"],
                },
            },
        },
        indent=2,
    )


class World:
    """A stand-in machine, a real supervisor over it, and a print to read."""

    def __init__(self, program: Path, root: Path) -> None:
        """Bring one up under `root`, running the program at `program`."""
        self.program = program
        self.root = root
        self.machine = Machine()
        self.state = root / "state"
        self.state.mkdir(parents=True, exist_ok=True)
        self._supervisor: subprocess.Popen[str] | None = None

    def __enter__(self) -> Running:
        """Start the supervisor and open a print for a client to read."""
        return self.start()

    def __exit__(
        self,
        kind: type[BaseException] | None,
        value: BaseException | None,
        traceback: TracebackType | None,
    ) -> None:
        """Stop everything this started."""
        self.stop()

    def start(self) -> Running:
        """Start the supervisor and open a print for a client to read.

        Raises:
            WorldError: If the supervisor did not come up, or opened no print.
        """
        configuration = self.root / "supervisor.toml"
        configuration.write_text(
            _as_toml(json.loads(_configuration(self.state, self.machine.url))),
            encoding="utf-8",
        )
        self._supervisor = start(
            [str(self.program), "server", "--config", str(configuration)],
            cwd=self.root,
        )
        server = self._await_address()
        print_id, image_id = self._open_a_print(server)
        return Running(server=server, print_id=print_id, image_id=image_id, state=self.state)

    def stop(self) -> None:
        """Stop the supervisor and the machine, leaving no process behind."""
        if self._supervisor is not None:
            self._supervisor.terminate()
            try:
                self._supervisor.wait(timeout=30)
            except subprocess.TimeoutExpired:
                self._supervisor.kill()
                self._supervisor.wait(timeout=30)
            self._supervisor = None
        self.machine.stop()

    def _await_address(self) -> str:
        """The address the supervisor bound, read from what it wrote for its clients.

        Raises:
            WorldError: If it never wrote one.
        """
        written = self.state / CLIENT_CONFIG
        deadline = time.monotonic() + STARTUP_TIMEOUT_SECONDS
        while time.monotonic() < deadline:
            if self._supervisor is not None and self._supervisor.poll() is not None:
                said = self._supervisor.communicate()[1]
                msg = f"the supervisor stopped before it answered:\n{said}"
                raise WorldError(msg)
            if written.is_file():
                for line in written.read_text(encoding="utf-8").splitlines():
                    name, _, value = line.partition("=")
                    if name.strip() == "server":
                        return value.strip().strip('"')
            time.sleep(0.1)
        msg = f"the supervisor wrote no {CLIENT_CONFIG} in {STARTUP_TIMEOUT_SECONDS}s"
        raise WorldError(msg)

    def _open_a_print(self, server: str) -> tuple[str, str]:
        """Open a print and an image by posting one alert to the ingress.

        Raises:
            WorldError: If the ingress opened none.
        """
        import http.client
        from urllib.parse import urlsplit

        alert = json.dumps(
            {
                "event": {"type": "PrintFailure", "is_warning": False, "print_paused": True},
                "printer": {"id": 17, "name": "A stand-in machine"},
                "print": {
                    "id": 4211,
                    "filename": "benchy.gcode",
                    "started_at": 1772366400.5,
                    "ended_at": "",
                },
                "img_url": f"{self.machine.url}/snapshot.jpg",
            }
        )
        connection = http.client.HTTPConnection(urlsplit(server).netloc, timeout=30)
        connection.request(
            "POST",
            "/obico/webhook",
            body=alert,
            headers={"Content-Type": "application/json", INGRESS_HEADER: INGRESS_WORD},
        )
        answered = connection.getresponse()
        answered.read()
        connection.close()
        # The ingress answers before its handling completes — it has a bound to
        # stay inside and a producer that does not retry — so what is asserted
        # here is that it took the body, and the print is waited for below.
        if answered.status not in ACCEPTED:
            msg = f"the ingress answered {answered.status} to the alert that opens a print"
            raise WorldError(msg)

        deadline = time.monotonic() + INGRESS_TIMEOUT_SECONDS
        while time.monotonic() < deadline:
            found = self._recorded()
            if found is not None:
                return found
            time.sleep(0.2)
        msg = "the ingress accepted the alert and no print and image were recorded"
        raise WorldError(msg)

    def _recorded(self) -> tuple[str, str] | None:
        """The print and image the ingress opened, read out of the store."""
        store = self.state / STORE
        if not store.is_file():
            return None
        connection = sqlite3.connect(f"file:{store}?mode=ro", uri=True)
        try:
            prints = connection.execute("SELECT id FROM prints").fetchall()
            images = connection.execute("SELECT id FROM images").fetchall()
        except sqlite3.DatabaseError:
            return None
        finally:
            connection.close()
        if not prints or not images:
            return None
        return str(prints[0][0]), str(images[0][0])


def _as_toml(document: dict[str, object]) -> str:
    """One configuration document, as the TOML the supervisor reads.

    Every value below the top level is written as an inline table, which is
    valid TOML whatever a key is spelled like — and one of them is spelled
    `tool_target:0`, which a section header cannot carry unquoted. Written here
    rather than with a serializer, because this repository takes no
    TOML-writing dependency.
    """
    return "".join(f"{_key(name)} = {_inline(value)}\n" for name, value in document.items())


def _key(name: str) -> str:
    """One key, quoted where TOML needs it to be."""
    plain = name.replace("_", "").replace("-", "").isalnum()
    return name if plain else json.dumps(name)


def _inline(value: object) -> str:
    """One value, as TOML writes it inline."""
    if isinstance(value, dict):
        written = ", ".join(f"{_key(str(k))} = {_inline(v)}" for k, v in value.items())
        return "{ " + written + " }"
    if isinstance(value, list):
        return "[" + ", ".join(_inline(entry) for entry in value) + "]"
    return json.dumps(value)
